File: LeetCode/DP1D/java/wordBreak.java
package LeetCode.DP1D.java;

import java.util.Arrays;
import java.util.List;

record InputVals(String s, List<String> wordDict) {
}

public class wordBreak {
    public boolean solveWordBreak(String s, List<String> wordDict) {
        int len_s = s.length();
        Boolean[] dp = new Boolean[len_s + 1];
        Arrays.fill(dp, false);

        dp[len_s] = true;

        for (int i = len_s - 1; i > -1; i--) {
            for (String w : wordDict) {
                if ((i + w.length()) <= len_s && (s.substring(i, i + w.length())).equals(w)) {
                    dp[i] = dp[i + w.length()];
                }
                if (dp[i] == true) {
                    break;
                }
            }
        }

        return dp[0];
    }

    public static void main(String[] args) {

        InputVals input = new InputVals("leetcode", List.of("leet", "code"));
        Boolean expected = Boolean.valueOf(true);
        Boolean result = new wordBreak().solveWordBreak(input.s(), input.wordDict());

        System.out.println("Input: " + input.toString());
        System.out.println("Result: " + result);
        System.out.println("Expected: " + expected);

        System.out.println("-".repeat(50));

        input = new InputVals("applepenapple", List.of("apple", "pen"));
        expected = Boolean.valueOf(true);
        result = new wordBreak().solveWordBreak(input.s(), input.wordDict());

        System.out.println("Input: " + input.toString());
        System.out.println("Result: " + result);
        System.out.println("Expected: " + expected);

        System.out.println("-".repeat(50));

        input = new InputVals("catsandog", List.of("cats", "dog", "sand", "and", "cat"));
        expected = Boolean.valueOf(false);
        result = new wordBreak().solveWordBreak(input.s(), input.wordDict());

        System.out.println("Input: " + input.toString());
        System.out.println("Result: " + result);
        System.out.println("Expected: " + expected);

    }
}

File: LeetCode/DP1D/python/wordBreak.py
"""Word break dynamic programming examples."""

from dataclasses import dataclass
from enum import Enum
from typing import List, Tuple


class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"


@dataclass
class WordBreakRecord:
    s: str
    wordDict: List[str]
    expected: bool


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[len(s)] = True

        for i in range(len(s) - 1, -1, -1):
            for word in wordDict:
                if i + len(word) <= len(s) and s[i : i + len(word)] == word:
                    dp[i] = dp[i + len(word)]
                if dp[i]:
                    break

        return dp[0]


def testSolution(record: WordBreakRecord) -> None:
    print(f"s: {record.s}")
    print(f"dict: {record.wordDict}")
    print(f"expected: {record.expected}")

    result = Solution().wordBreak(record.s, record.wordDict)

    print(f"result: {result}")
    print(Result.PASS.value if result == record.expected else Result.FAIL.value)


if __name__ == "__main__":
    records: Tuple[WordBreakRecord, ...] = (
        WordBreakRecord("leetcode", ["leet", "code"], True),
        WordBreakRecord("applepenapple", ["apple", "pen"], True),
        WordBreakRecord("catsandog", ["cats", "dog", "sand", "and", "cat"], False),
    )

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        testSolution(record)
        print("-" * 50)
